File: Leetcode 2001-2100/2055/2055.java
// Leetcode 2055: Plates Between Candles
// https://leetcode.com/problems/plates-between-candles/
// Solved on 7th of November, 2025
class Solution {
    /**
     * Calculates the number of plates between candles for given queries.
     * @param s The string representing plates and candles.
     * @param queries An array of queries, where each query is [start, end].
     * @return An array of integers, where each element is the number of plates between candles for the corresponding query.
     */
    public int[] platesBetweenCandles(String s, int[][] queries) {
        int n = s.length();
        char[] arr = s.toCharArray();
        int[] prefix = new int[n];
        int count = 0;
        for (int i = 0; i < n; i++) {
            if (arr[i] == '*') {
                count++;
            }
            prefix[i] = count;
        }
        int[] nearestLeft = new int[n];
        int last = -1;
        for (int i = 0; i < n; i++) {
            if (arr[i] == '|') {
                last = i;
            }
            nearestLeft[i] = last;
        }
        int[] nearestRight = new int[n];
        last = -1;
        for (int i = n - 1; i >= 0; i--) {
            if (arr[i] == '|') {
                last = i;
            }
            nearestRight[i] = last;
        }
        int m = queries.length;
        int[] result = new int[m];
        for (int i = 0; i < m; i++) {
            int left = queries[i][0];
            int right = queries[i][1];
            int leftCandle = nearestRight[left];
            int rightCandle = nearestLeft[right];
            if (leftCandle == -1 || rightCandle == -1 || leftCandle >= rightCandle) {
                result[i] = 0;
            } else {
                result[i] = prefix[rightCandle] - prefix[leftCandle];
            }
        }
        return result;
    }
}

File: Leetcode 2001-2100/2055/2055.py
# Leetcode 2055: Plates Between Candles
# https://leetcode.com/problems/plates-between-candles/
# Solved on 7th of November, 2025
class Solution:
    def platesBetweenCandles(self, s: str, queries: list[list[int]]) -> list[int]:
        """
        Calculates the number of plates between candles for given queries.

        Args:
            s (str): A string representing plates '*' and candles '|'.
            queries (list[list[int]]): A list of query ranges [left, right].
        Returns:
            list[int]: A list of integers, where each element is the number of plates
                       between candles for the corresponding query.
        """

        n = len(s)

        prefixPlates = [0] * (n + 1)
        for i in range(n):
            prefixPlates[i + 1] = prefixPlates[i]
            if s[i] == '*':
                prefixPlates[i + 1] += 1

        prevCandle = [-1] * n
        lastSeenCandle = -1
        for i in range(n):
            if s[i] == '|':
                lastSeenCandle = i
            prevCandle[i] = lastSeenCandle

        nextCandle = [-1] * n
        lastSeenCandle = -1
        for i in range(n - 1, -1, -1):
            if s[i] == '|':
                lastSeenCandle = i
            nextCandle[i] = lastSeenCandle

        results = []
        for query in queries:
            qLeft = query[0]
            qRight = query[1]

            firstCandle = nextCandle[qLeft]
            lastCandle = prevCandle[qRight]

            if firstCandle == -1 or lastCandle == -1 or firstCandle >= lastCandle:
                results.append(0)
            else:
                plateCount = prefixPlates[lastCandle + 1] - prefixPlates[firstCandle]
                results.append(plateCount)

        return results